ic-canister-client/src/pocket_ic.rs

```diff
@@ -0,0 +1,160 @@
+use std::sync::Arc;
+
+use candid::utils::ArgumentEncoder;
+use candid::{CandidType, Decode, Principal};
+use ic_exports::ic_kit::RejectionCode;
+use pocket_ic::{PocketIc, WasmResult};
+use serde::Deserialize;
+
+use crate::{CanisterClient, CanisterClientError, CanisterClientResult};
+
+/// A client for interacting with a canister inside dfinity's PocketIc test framework.
+#[derive(Clone)]
+pub struct PocketIcClient {
+    client: Arc<PocketIc>,
+    canister: Principal,
+    caller: Principal,
+}
+
+impl PocketIcClient {
+    /// Creates a new instance of a PocketIcClient.
+    /// The new instance is independent and have no access to canisters of other instances.
+    pub fn new(canister: Principal, caller: Principal) -> Self {
+        Self {
+            client: Arc::new(PocketIc::new()),
+            canister,
+            caller,
+        }
+    }
+
+    /// Crates new instance of PocketIcClient from an existing client instance.
+    pub fn from_client(client: Arc<PocketIc>, canister: Principal, caller: Principal) -> Self {
+        Self {
+            client,
+            canister,
+            caller,
+        }
+    }
+
+    /// Returns the caller of the canister.
+    pub fn caller(&self) -> Principal {
+        self.caller
+    }
+
+    /// Replace the caller.
+    pub fn set_caller(&mut self, caller: Principal) {
```

Comment:
This looks a lot like Java getter/setter. Why don't you make the field public instead?

Reply:
Done